presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java

```diff
@@ -390,7 +390,17 @@ protected ExchangeNode build()
 
     public JoinNode join(JoinNode.Type joinType, PlanNode left, PlanNode right, JoinNode.EquiJoinClause... criteria)
     {
-        return new JoinNode(idAllocator.getNextId(),
+        return join(joinType, left, right, Optional.empty(), criteria);
+    }
+
+    public JoinNode join(JoinNode.Type joinType, PlanNode left, PlanNode right, Expression filter, JoinNode.EquiJoinClause... criteria)
+    {
+        return join(joinType, left, right, Optional.of(filter), criteria);
+    }
+
+    private JoinNode join(JoinNode.Type joinType, PlanNode left, PlanNode right, Optional<Expression> filter, JoinNode.EquiJoinClause... criteria)
```

Comment:
It would be more natural to have `filter` after `criteria` (cause criteria are considered first and are first in `JoinNode` ctor), but the cost of prohibiting varargs is too high. Did we cross the line where builder would be appropriate?